// roomjs/dist/room.js
(function(f) {
    if (typeof exports === "object" && typeof module !== "undefined") {
        module.exports = f()
    } else if (typeof define === "function" && define.amd) {
        define([], f)
    } else {
        var g;
        if (typeof window !== "undefined") {
            g = window
        } else if (typeof global !== "undefined") {
            g = global
        } else if (typeof self !== "undefined") {
            g = self
        } else {
            g = this
        }
        g.Room = f()
    }
})(function() {
    var define, module, exports;
    return (function e(t, n, r) {
        function s(o, u) {
            if (!n[o]) {
                if (!t[o]) {
                    var a = typeof require == "function" && require;
                    if (!u && a) return a(o, !0);
                    if (i) return i(o, !0);
                    var f = new Error("Cannot find module '" + o + "'");
                    throw f.code = "MODULE_NOT_FOUND", f
                }
                var l = n[o] = {
                    exports: {}
                };
                t[o][0].call(l.exports, function(e) {
                    var n = t[o][1][e];
                    return s(n ? n : e)
                }, l, l.exports, e, t, n, r)
            }
            return n[o].exports
        }
        var i = typeof require == "function" && require;
        for (var o = 0; o < r.length; o++) s(r[o]);
        return s
    })({
        1: [function(require, module, exports) {
            // shim for using process in browser

            var process = module.exports = {};
            var queue = [];
            var draining = false;
            var currentQueue;
            var queueIndex = -1;

            function cleanUpNextTick() {
                draining = false;
                if (currentQueue.length) {
                    queue = currentQueue.concat(queue);
                } else {
                    queueIndex = -1;
                }
                if (queue.length) {
                    drainQueue();
                }
            }

            function drainQueue() {
                if (draining) {
                    return;
                }
                var timeout = setTimeout(cleanUpNextTick);
                draining = true;

                var len = queue.length;
                while (len) {
                    currentQueue = queue;
                    queue = [];
                    while (++queueIndex < len) {
                        if (currentQueue) {
                            currentQueue[queueIndex].run();
                        }
                    }
                    queueIndex = -1;
                    len = queue.length;
                }
                currentQueue = null;
                draining = false;
                clearTimeout(timeout);
            }

            process.nextTick = function(fun) {
                var args = new Array(arguments.length - 1);
                if (arguments.length > 1) {
                    for (var i = 1; i < arguments.length; i++) {
                        args[i - 1] = arguments[i];
                    }
                }
                queue.push(new Item(fun, args));
                if (queue.length === 1 && !draining) {
                    setTimeout(drainQueue, 0);
                }
            };

            // v8 likes predictible objects
            function Item(fun, array) {
                this.fun = fun;
                this.array = array;
            }
            Item.prototype.run = function() {
                this.fun.apply(null, this.array);
            };
            process.title = 'browser';
            process.browser = true;
            process.env = {};
            process.argv = [];
            process.version = ''; // empty string to avoid regexp issues
            process.versions = {};

            function noop() {}

            process.on = noop;
            process.addListener = noop;
            process.once = noop;
            process.off = noop;
            process.removeListener = noop;
            process.removeAllListeners = noop;
            process.emit = noop;

            process.binding = function(name) {
                throw new Error('process.binding is not supported');
            };

            process.cwd = function() {
                return '/'
            };
            process.chdir = function(dir) {
                throw new Error('process.chdir is not supported');
            };
            process.umask = function() {
                return 0;
            };

        }, {}],
        2: [function(require, module, exports) {
            (function(process) {
                'use strict';
                var escapeStringRegexp = require('escape-string-regexp');
                var ansiStyles = require('ansi-styles');
                var stripAnsi = require('strip-ansi');
                var hasAnsi = require('has-ansi');
                var supportsColor = require('supports-color');
                var defineProps = Object.defineProperties;
                var isSimpleWindowsTerm = process.platform === 'win32' && !/^xterm/i.test(process.env.TERM);

                function Chalk(options) {
                    // detect mode if not set manually
                    this.enabled = !options || options.enabled === undefined ? supportsColor : options.enabled;
                }

                // use bright blue on Windows as the normal blue color is illegible
                if (isSimpleWindowsTerm) {
                    ansiStyles.blue.open = '\u001b[94m';
                }

                var styles = (function() {
                    var ret = {};

                    Object.keys(ansiStyles).forEach(function(key) {
                        ansiStyles[key].closeRe = new RegExp(escapeStringRegexp(ansiStyles[key].close), 'g');

                        ret[key] = {
                            get: function() {
                                return build.call(this, this._styles.concat(key));
                            }
                        };
                    });

                    return ret;
                })();

                var proto = defineProps(function chalk() {}, styles);

                function build(_styles) {
                    var builder = function() {
                        return applyStyle.apply(builder, arguments);
                    };

                    builder._styles = _styles;
                    builder.enabled = this.enabled;
                    // __proto__ is used because we must return a function, but there is
                    // no way to create a function with a different prototype.
                    /* eslint-disable no-proto */
                    builder.__proto__ = proto;

                    return builder;
                }

                function applyStyle() {
                    // support varags, but simply cast to string in case there's only one arg
                    var args = arguments;
                    var argsLen = args.length;
                    var str = argsLen !== 0 && String(arguments[0]);

                    if (argsLen > 1) {
                        // don't slice `arguments`, it prevents v8 optimizations
                        for (var a = 1; a < argsLen; a++) {
                            str += ' ' + args[a];
                        }
                    }

                    if (!this.enabled || !str) {
                        return str;
                    }

                    var nestedStyles = this._styles;
                    var i = nestedStyles.length;

                    // Turns out that on Windows dimmed gray text becomes invisible in cmd.exe,
                    // see https://github.com/chalk/chalk/issues/58
                    // If we're on Windows and we're dealing with a gray color, temporarily make 'dim' a noop.
                    var originalDim = ansiStyles.dim.open;
                    if (isSimpleWindowsTerm && (nestedStyles.indexOf('gray') !== -1 || nestedStyles.indexOf('grey') !== -1)) {
                        ansiStyles.dim.open = '';
                    }

                    while (i--) {
                        var code = ansiStyles[nestedStyles[i]];

                        // Replace any instances already present with a re-opening code
                        // otherwise only the part of the string until said closing code
                        // will be colored, and the rest will simply be 'plain'.
                        str = code.open + str.replace(code.closeRe, code.open) + code.close;
                    }

                    // Reset the original 'dim' if we changed it to work around the Windows dimmed gray issue.
                    ansiStyles.dim.open = originalDim;

                    return str;
                }

                function init() {
                    var ret = {};

                    Object.keys(styles).forEach(function(name) {
                        ret[name] = {
                            get: function() {
                                return build.call(this, [name]);
                            }
                        };
                    });

                    return ret;
                }

                defineProps(Chalk.prototype, init());

                module.exports = new Chalk();
                module.exports.styles = ansiStyles;
                module.exports.hasColor = hasAnsi;
                module.exports.stripColor = stripAnsi;
                module.exports.supportsColor = supportsColor;

            }).call(this, require('_process'))

        }, {
            "_process": 1,
            "ansi-styles": 3,
            "escape-string-regexp": 4,
            "has-ansi": 5,
            "strip-ansi": 7,
            "supports-color": 9
        }],
        3: [function(require, module, exports) {
            'use strict';

            function assembleStyles() {
                var styles = {
                    modifiers: {
                        reset: [0, 0],
                        bold: [1, 22], // 21 isn't widely supported and 22 does the same thing
                        dim: [2, 22],
                        italic: [3, 23],
                        underline: [4, 24],
                        inverse: [7, 27],
                        hidden: [8, 28],
                        strikethrough: [9, 29]
                    },
                    colors: {
                        black: [30, 39],
                        red: [31, 39],
                        green: [32, 39],
                        yellow: [33, 39],
                        blue: [34, 39],
                        magenta: [35, 39],
                        cyan: [36, 39],
                        white: [37, 39],
                        gray: [90, 39]
                    },
                    bgColors: {
                        bgBlack: [40, 49],
                        bgRed: [41, 49],
                        bgGreen: [42, 49],
                        bgYellow: [43, 49],
                        bgBlue: [44, 49],
                        bgMagenta: [45, 49],
                        bgCyan: [46, 49],
                        bgWhite: [47, 49]
                    }
                };

                // fix humans
                styles.colors.grey = styles.colors.gray;

                Object.keys(styles).forEach(function(groupName) {
                    var group = styles[groupName];

                    Object.keys(group).forEach(function(styleName) {
                        var style = group[styleName];

                        styles[styleName] = group[styleName] = {
                            open: '\u001b[' + style[0] + 'm',
                            close: '\u001b[' + style[1] + 'm'
                        };
                    });

                    Object.defineProperty(styles, groupName, {
                        value: group,
                        enumerable: false
                    });
                });

                return styles;
            }

            Object.defineProperty(module, 'exports', {
                enumerable: true,
                get: assembleStyles
            });

        }, {}],
        4: [function(require, module, exports) {
            'use strict';

            var matchOperatorsRe = /[|\\{}()[\]^$+*?.]/g;

            module.exports = function(str) {
                if (typeof str !== 'string') {
                    throw new TypeError('Expected a string');
                }

                return str.replace(matchOperatorsRe, '\\$&');
            };

        }, {}],
        5: [function(require, module, exports) {
            'use strict';
            var ansiRegex = require('ansi-regex');
            var re = new RegExp(ansiRegex().source); // remove the `g` flag
            module.exports = re.test.bind(re);

        }, {
            "ansi-regex": 6
        }],
        6: [function(require, module, exports) {
            'use strict';
            module.exports = function() {
                return /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g;
            };

        }, {}],
        7: [function(require, module, exports) {
            'use strict';
            var ansiRegex = require('ansi-regex')();

            module.exports = function(str) {
                return typeof str === 'string' ? str.replace(ansiRegex, '') : str;
            };

        }, {
            "ansi-regex": 8
        }],
        8: [function(require, module, exports) {
            arguments[4][6][0].apply(exports, arguments)
        }, {
            "dup": 6
        }],
        9: [function(require, module, exports) {
            (function(process) {
                'use strict';
                var argv = process.argv;

                var terminator = argv.indexOf('--');
                var hasFlag = function(flag) {
                    flag = '--' + flag;
                    var pos = argv.indexOf(flag);
                    return pos !== -1 && (terminator !== -1 ? pos < terminator : true);
                };

                module.exports = (function() {
                    if ('FORCE_COLOR' in process.env) {
                        return true;
                    }

                    if (hasFlag('no-color') ||
                        hasFlag('no-colors') ||
                        hasFlag('color=false')) {
                        return false;
                    }

                    if (hasFlag('color') ||
                        hasFlag('colors') ||
                        hasFlag('color=true') ||
                        hasFlag('color=always')) {
                        return true;
                    }

                    if (process.stdout && !process.stdout.isTTY) {
                        return false;
                    }

                    if (process.platform === 'win32') {
                        return true;
                    }

                    if ('COLORTERM' in process.env) {
                        return true;
                    }

                    if (process.env.TERM === 'dumb') {
                        return false;
                    }

                    if (/^screen|^xterm|^vt100|color|ansi|cygwin|linux/i.test(process.env.TERM)) {
                        return true;
                    }

                    return false;
                })();

            }).call(this, require('_process'))

        }, {
            "_process": 1
        }],
        10: [function(require, module, exports) {
            /*!
             * mustache.js - Logic-less {{mustache}} templates with JavaScript
             * http://github.com/janl/mustache.js
             */

            /*global define: false Mustache: true*/

            (function defineMustache(global, factory) {
                if (typeof exports === 'object' && exports && typeof exports.nodeName !== 'string') {
                    factory(exports); // CommonJS
                } else if (typeof define === 'function' && define.amd) {
                    define(['exports'], factory); // AMD
                } else {
                    global.Mustache = {};
                    factory(Mustache); // script, wsh, asp
                }
            }(this, function mustacheFactory(mustache) {

                var objectToString = Object.prototype.toString;
                var isArray = Array.isArray || function isArrayPolyfill(object) {
                    return objectToString.call(object) === '[object Array]';
                };

                function isFunction(object) {
                    return typeof object === 'function';
                }

                /**
                 * More correct typeof string handling array
                 * which normally returns typeof 'object'
                 */
                function typeStr(obj) {
                    return isArray(obj) ? 'array' : typeof obj;
                }

                function escapeRegExp(string) {
                    return string.replace(/[\-\[\]{}()*+?.,\\\^$|#\s]/g, '\\$&');
                }

                /**
                 * Null safe way of checking whether or not an object,
                 * including its prototype, has a given property
                 */
                function hasProperty(obj, propName) {
                    return obj != null && typeof obj === 'object' && (propName in obj);
                }

                // Workaround for https://issues.apache.org/jira/browse/COUCHDB-577
                // See https://github.com/janl/mustache.js/issues/189
                var regExpTest = RegExp.prototype.test;

                function testRegExp(re, string) {
                    return regExpTest.call(re, string);
                }

                var nonSpaceRe = /\S/;

                function isWhitespace(string) {
                    return !testRegExp(nonSpaceRe, string);
                }

                var entityMap = {
                    '&': '&amp;',
                    '<': '&lt;',
                    '>': '&gt;',
                    '"': '&quot;',
                    "'": '&#39;',
                    '/': '&#x2F;'
                };

                function escapeHtml(string) {
                    return String(string).replace(/[&<>"'\/]/g, function fromEntityMap(s) {
                        return entityMap[s];
                    });
                }

                var whiteRe = /\s*/;
                var spaceRe = /\s+/;
                var equalsRe = /\s*=/;
                var curlyRe = /\s*\}/;
                var tagRe = /#|\^|\/|>|\{|&|=|!/;

                /**
                 * Breaks up the given `template` string into a tree of tokens. If the `tags`
                 * argument is given here it must be an array with two string values: the
                 * opening and closing tags used in the template (e.g. [ "<%", "%>" ]). Of
                 * course, the default is to use mustaches (i.e. mustache.tags).
                 *
                 * A token is an array with at least 4 elements. The first element is the
                 * mustache symbol that was used inside the tag, e.g. "#" or "&". If the tag
                 * did not contain a symbol (i.e. {{myValue}}) this element is "name". For
                 * all text that appears outside a symbol this element is "text".
                 *
                 * The second element of a token is its "value". For mustache tags this is
                 * whatever else was inside the tag besides the opening symbol. For text tokens
                 * this is the text itself.
                 *
                 * The third and fourth elements of the token are the start and end indices,
                 * respectively, of the token in the original template.
                 *
                 * Tokens that are the root node of a subtree contain two more elements: 1) an
                 * array of tokens in the subtree and 2) the index in the original template at
                 * which the closing tag for that section begins.
                 */
                function parseTemplate(template, tags) {
                    if (!template)
                        return [];

                    var sections = []; // Stack to hold section tokens
                    var tokens = []; // Buffer to hold the tokens
                    var spaces = []; // Indices of whitespace tokens on the current line
                    var hasTag = false; // Is there a {{tag}} on the current line?
                    var nonSpace = false; // Is there a non-space char on the current line?

                    // Strips all whitespace tokens array for the current line
                    // if there was a {{#tag}} on it and otherwise only space.
                    function stripSpace() {
                        if (hasTag && !nonSpace) {
                            while (spaces.length)
                                delete tokens[spaces.pop()];
                        } else {
                            spaces = [];
                        }

                        hasTag = false;
                        nonSpace = false;
                    }

                    var openingTagRe, closingTagRe, closingCurlyRe;

                    function compileTags(tagsToCompile) {
                        if (typeof tagsToCompile === 'string')
                            tagsToCompile = tagsToCompile.split(spaceRe, 2);

                        if (!isArray(tagsToCompile) || tagsToCompile.length !== 2)
                            throw new Error('Invalid tags: ' + tagsToCompile);

                        openingTagRe = new RegExp(escapeRegExp(tagsToCompile[0]) + '\\s*');
                        closingTagRe = new RegExp('\\s*' + escapeRegExp(tagsToCompile[1]));
                        closingCurlyRe = new RegExp('\\s*' + escapeRegExp('}' + tagsToCompile[1]));
                    }

                    compileTags(tags || mustache.tags);

                    var scanner = new Scanner(template);

                    var start, type, value, chr, token, openSection;
                    while (!scanner.eos()) {
                        start = scanner.pos;

                        // Match any text between tags.
                        value = scanner.scanUntil(openingTagRe);

                        if (value) {
                            for (var i = 0, valueLength = value.length; i < valueLength; ++i) {
                                chr = value.charAt(i);

                                if (isWhitespace(chr)) {
                                    spaces.push(tokens.length);
                                } else {
                                    nonSpace = true;
                                }

                                tokens.push(['text', chr, start, start + 1]);
                                start += 1;

                                // Check for whitespace on the current line.
                                if (chr === '\n')
                                    stripSpace();
                            }
                        }

                        // Match the opening tag.
                        if (!scanner.scan(openingTagRe))
                            break;

                        hasTag = true;

                        // Get the tag type.
                        type = scanner.scan(tagRe) || 'name';
                        scanner.scan(whiteRe);

                        // Get the tag value.
                        if (type === '=') {
                            value = scanner.scanUntil(equalsRe);
                            scanner.scan(equalsRe);
                            scanner.scanUntil(closingTagRe);
                        } else if (type === '{') {
                            value = scanner.scanUntil(closingCurlyRe);
                            scanner.scan(curlyRe);
                            scanner.scanUntil(closingTagRe);
                            type = '&';
                        } else {
                            value = scanner.scanUntil(closingTagRe);
                        }

                        // Match the closing tag.
                        if (!scanner.scan(closingTagRe))
                            throw new Error('Unclosed tag at ' + scanner.pos);

                        token = [type, value, start, scanner.pos];
                        tokens.push(token);

                        if (type === '#' || type === '^') {
                            sections.push(token);
                        } else if (type === '/') {
                            // Check section nesting.
                            openSection = sections.pop();

                            if (!openSection)
                                throw new Error('Unopened section "' + value + '" at ' + start);

                            if (openSection[1] !== value)
                                throw new Error('Unclosed section "' + openSection[1] + '" at ' + start);
                        } else if (type === 'name' || type === '{' || type === '&') {
                            nonSpace = true;
                        } else if (type === '=') {
                            // Set the tags for the next time around.
                            compileTags(value);
                        }
                    }

                    // Make sure there are no open sections when we're done.
                    openSection = sections.pop();

                    if (openSection)
                        throw new Error('Unclosed section "' + openSection[1] + '" at ' + scanner.pos);

                    return nestTokens(squashTokens(tokens));
                }

                /**
                 * Combines the values of consecutive text tokens in the given `tokens` array
                 * to a single token.
                 */
                function squashTokens(tokens) {
                    var squashedTokens = [];

                    var token, lastToken;
                    for (var i = 0, numTokens = tokens.length; i < numTokens; ++i) {
                        token = tokens[i];

                        if (token) {
                            if (token[0] === 'text' && lastToken && lastToken[0] === 'text') {
                                lastToken[1] += token[1];
                                lastToken[3] = token[3];
                            } else {
                                squashedTokens.push(token);
                                lastToken = token;
                            }
                        }
                    }

                    return squashedTokens;
                }

                /**
                 * Forms the given array of `tokens` into a nested tree structure where
                 * tokens that represent a section have two additional items: 1) an array of
                 * all tokens that appear in that section and 2) the index in the original
                 * template that represents the end of that section.
                 */
                function nestTokens(tokens) {
                    var nestedTokens = [];
                    var collector = nestedTokens;
                    var sections = [];

                    var token, section;
                    for (var i = 0, numTokens = tokens.length; i < numTokens; ++i) {
                        token = tokens[i];

                        switch (token[0]) {
                            case '#':
                            case '^':
                                collector.push(token);
                                sections.push(token);
                                collector = token[4] = [];
                                break;
                            case '/':
                                section = sections.pop();
                                section[5] = token[2];
                                collector = sections.length > 0 ? sections[sections.length - 1][4] : nestedTokens;
                                break;
                            default:
                                collector.push(token);
                        }
                    }

                    return nestedTokens;
                }

                /**
                 * A simple string scanner that is used by the template parser to find
                 * tokens in template strings.
                 */
                function Scanner(string) {
                    this.string = string;
                    this.tail = string;
                    this.pos = 0;
                }

                /**
                 * Returns `true` if the tail is empty (end of string).
                 */
                Scanner.prototype.eos = function eos() {
                    return this.tail === '';
                };

                /**
                 * Tries to match the given regular expression at the current position.
                 * Returns the matched text if it can match, the empty string otherwise.
                 */
                Scanner.prototype.scan = function scan(re) {
                    var match = this.tail.match(re);

                    if (!match || match.index !== 0)
                        return '';

                    var string = match[0];

                    this.tail = this.tail.substring(string.length);
                    this.pos += string.length;

                    return string;
                };

                /**
                 * Skips all text until the given regular expression can be matched. Returns
                 * the skipped string, which is the entire tail if no match can be made.
                 */
                Scanner.prototype.scanUntil = function scanUntil(re) {
                    var index = this.tail.search(re),
                        match;

                    switch (index) {
                        case -1:
                            match = this.tail;
                            this.tail = '';
                            break;
                        case 0:
                            match = '';
                            break;
                        default:
                            match = this.tail.substring(0, index);
                            this.tail = this.tail.substring(index);
                    }

                    this.pos += match.length;

                    return match;
                };

                /**
                 * Represents a rendering context by wrapping a view object and
                 * maintaining a reference to the parent context.
                 */
                function Context(view, parentContext) {
                    this.view = view;
                    this.cache = {
                        '.': this.view
                    };
                    this.parent = parentContext;
                }

                /**
                 * Creates a new context using the given view with this context
                 * as the parent.
                 */
                Context.prototype.push = function push(view) {
                    return new Context(view, this);
                };

                /**
                 * Returns the value of the given name in this context, traversing
                 * up the context hierarchy if the value is absent in this context's view.
                 */
                Context.prototype.lookup = function lookup(name) {
                    var cache = this.cache;

                    var value;
                    if (cache.hasOwnProperty(name)) {
                        value = cache[name];
                    } else {
                        var context = this,
                            names, index, lookupHit = false;

                        while (context) {
                            if (name.indexOf('.') > 0) {
                                value = context.view;
                                names = name.split('.');
                                index = 0;

                                /**
                                 * Using the dot notion path in `name`, we descend through the
                                 * nested objects.
                                 *
                                 * To be certain that the lookup has been successful, we have to
                                 * check if the last object in the path actually has the property
                                 * we are looking for. We store the result in `lookupHit`.
                                 *
                                 * This is specially necessary for when the value has been set to
                                 * `undefined` and we want to avoid looking up parent contexts.
                                 **/
                                while (value != null && index < names.length) {
                                    if (index === names.length - 1)
                                        lookupHit = hasProperty(value, names[index]);

                                    value = value[names[index++]];
                                }
                            } else {
                                value = context.view[name];
                                lookupHit = hasProperty(context.view, name);
                            }

                            if (lookupHit)
                                break;

                            context = context.parent;
                        }

                        cache[name] = value;
                    }

                    if (isFunction(value))
                        value = value.call(this.view);

                    return value;
                };

                /**
                 * A Writer knows how to take a stream of tokens and render them to a
                 * string, given a context. It also maintains a cache of templates to
                 * avoid the need to parse the same template twice.
                 */
                function Writer() {
                    this.cache = {};
                }

                /**
                 * Clears all cached templates in this writer.
                 */
                Writer.prototype.clearCache = function clearCache() {
                    this.cache = {};
                };

                /**
                 * Parses and caches the given `template` and returns the array of tokens
                 * that is generated from the parse.
                 */
                Writer.prototype.parse = function parse(template, tags) {
                    var cache = this.cache;
                    var tokens = cache[template];

                    if (tokens == null)
                        tokens = cache[template] = parseTemplate(template, tags);

                    return tokens;
                };

                /**
                 * High-level method that is used to render the given `template` with
                 * the given `view`.
                 *
                 * The optional `partials` argument may be an object that contains the
                 * names and templates of partials that are used in the template. It may
                 * also be a function that is used to load partial templates on the fly
                 * that takes a single argument: the name of the partial.
                 */
                Writer.prototype.render = function render(template, view, partials) {
                    var tokens = this.parse(template);
                    var context = (view instanceof Context) ? view : new Context(view);
                    return this.renderTokens(tokens, context, partials, template);
                };

                /**
                 * Low-level method that renders the given array of `tokens` using
                 * the given `context` and `partials`.
                 *
                 * Note: The `originalTemplate` is only ever used to extract the portion
                 * of the original template that was contained in a higher-order section.
                 * If the template doesn't use higher-order sections, this argument may
                 * be omitted.
                 */
                Writer.prototype.renderTokens = function renderTokens(tokens, context, partials, originalTemplate) {
                    var buffer = '';

                    var token, symbol, value;
                    for (var i = 0, numTokens = tokens.length; i < numTokens; ++i) {
                        value = undefined;
                        token = tokens[i];
                        symbol = token[0];

                        if (symbol === '#') value = this.renderSection(token, context, partials, originalTemplate);
                        else if (symbol === '^') value = this.renderInverted(token, context, partials, originalTemplate);
                        else if (symbol === '>') value = this.renderPartial(token, context, partials, originalTemplate);
                        else if (symbol === '&') value = this.unescapedValue(token, context);
                        else if (symbol === 'name') value = this.escapedValue(token, context);
                        else if (symbol === 'text') value = this.rawValue(token);

                        if (value !== undefined)
                            buffer += value;
                    }

                    return buffer;
                };

                Writer.prototype.renderSection = function renderSection(token, context, partials, originalTemplate) {
                    var self = this;
                    var buffer = '';
                    var value = context.lookup(token[1]);

                    // This function is used to render an arbitrary template
                    // in the current context by higher-order sections.
                    function subRender(template) {
                        return self.render(template, context, partials);
                    }

                    if (!value) return;

                    if (isArray(value)) {
                        for (var j = 0, valueLength = value.length; j < valueLength; ++j) {
                            buffer += this.renderTokens(token[4], context.push(value[j]), partials, originalTemplate);
                        }
                    } else if (typeof value === 'object' || typeof value === 'string' || typeof value === 'number') {
                        buffer += this.renderTokens(token[4], context.push(value), partials, originalTemplate);
                    } else if (isFunction(value)) {
                        if (typeof originalTemplate !== 'string')
                            throw new Error('Cannot use higher-order sections without the original template');

                        // Extract the portion of the original template that the section contains.
                        value = value.call(context.view, originalTemplate.slice(token[3], token[5]), subRender);

                        if (value != null)
                            buffer += value;
                    } else {
                        buffer += this.renderTokens(token[4], context, partials, originalTemplate);
                    }
                    return buffer;
                };

                Writer.prototype.renderInverted = function renderInverted(token, context, partials, originalTemplate) {
                    var value = context.lookup(token[1]);

                    // Use JavaScript's definition of falsy. Include empty arrays.
                    // See https://github.com/janl/mustache.js/issues/186
                    if (!value || (isArray(value) && value.length === 0))
                        return this.renderTokens(token[4], context, partials, originalTemplate);
                };

                Writer.prototype.renderPartial = function renderPartial(token, context, partials) {
                    if (!partials) return;

                    var value = isFunction(partials) ? partials(token[1]) : partials[token[1]];
                    if (value != null)
                        return this.renderTokens(this.parse(value), context, partials, value);
                };

                Writer.prototype.unescapedValue = function unescapedValue(token, context) {
                    var value = context.lookup(token[1]);
                    if (value != null)
                        return value;
                };

                Writer.prototype.escapedValue = function escapedValue(token, context) {
                    var value = context.lookup(token[1]);
                    if (value != null)
                        return mustache.escape(value);
                };

                Writer.prototype.rawValue = function rawValue(token) {
                    return token[1];
                };

                mustache.name = 'mustache.js';
                mustache.version = '2.1.3';
                mustache.tags = ['{{', '}}'];

                // All high-level mustache.* functions use this writer.
                var defaultWriter = new Writer();

                /**
                 * Clears all cached templates in the default writer.
                 */
                mustache.clearCache = function clearCache() {
                    return defaultWriter.clearCache();
                };

                /**
                 * Parses and caches the given template in the default writer and returns the
                 * array of tokens it contains. Doing this ahead of time avoids the need to
                 * parse templates on the fly as they are rendered.
                 */
                mustache.parse = function parse(template, tags) {
                    return defaultWriter.parse(template, tags);
                };

                /**
                 * Renders the `template` with the given `view` and `partials` using the
                 * default writer.
                 */
                mustache.render = function render(template, view, partials) {
                    if (typeof template !== 'string') {
                        throw new TypeError('Invalid template! Template should be a "string" ' +
                            'but "' + typeStr(template) + '" was given as the first ' +
                            'argument for mustache#render(template, view, partials)');
                    }

                    return defaultWriter.render(template, view, partials);
                };

                // This is here for backwards compatibility with 0.4.x.,
                /*eslint-disable */ // eslint wants camel cased function name
                mustache.to_html = function to_html(template, view, partials, send) {
                    /*eslint-enable*/

                    var result = mustache.render(template, view, partials);

                    if (isFunction(send)) {
                        send(result);
                    } else {
                        return result;
                    }
                };

                // Export the escaping function so that the user may override it.
                // See https://github.com/janl/mustache.js/issues/244
                mustache.escape = escapeHtml;

                // Export these mainly for testing, but also for advanced usage.
                mustache.Scanner = Scanner;
                mustache.Context = Context;
                mustache.Writer = Writer;

            }));

        }, {}],
        11: [function(require, module, exports) {
            var util = require('./util');
            var compiler = require('./compiler');
            var inject = require('./dependency').inject;
            var element = require('./element');
            var appfactory = require('./factory');
            var debug = util.debug('Room.application', 0);

            function RoomApplication(name, deps, opts) {
                this.name = name;
                this.components = {};
                this.services = {};
                this.elements = {};

                if (!Array.isArray(deps) && typeof deps === 'object') {
                    opts = deps;
                    deps = [];
                }

                this.dependencies = deps ||  [];
                this.options = opts ||  {};

                this.router = {
                    url: window.location.pathname,
                    filename: function(url) {
                        return url.substring(url.lastIndexOf('/') + 1);
                    }
                };

                // Resolve dependencies for application module
                this.checkDependencies();
            };

            RoomApplication.prototype.checkDependencies = function() {
                var self = this;
                debug.start('checkDependencies');
                self.dependencies.forEach(function(dep) {
                    if (!window[dep]) {
                        debug.error(
                            'Dependency ' + dep + ' is required for using ' +
                            'module \'' + self.name + '\''
                        );
                        return;
                    }
                });
                debug.info('No dependency found or everything included');
                debug.end();
            };

            RoomApplication.prototype.register = function(name, factory) {
                debug.start('register');
                var self = this;
                var handler, deps, injector = [],
                    required = [],
                    ready = true;
                if (Array.isArray(factory)) {
                    handler = factory.pop();
                    deps = factory;
                } else {
                    handler = factory;
                    deps = [];
                }

                deps.forEach(function(needed) {
                    var searching = inject(self, needed);
                    if (!searching) {
                        ready = false;
                        required.push(needed);
                    } else {
                        injector.push(searching);
                    }
                });

                if (!(self.components[name])) {
                    var internal = {
                        name: name,
                        ready: ready,
                        dependencies: deps,
                        inject: injector,
                        required: required
                    };

                    internal = appfactory.component(internal);
                    self.components[name] = internal;
                    internal.factory = handler.apply(internal, internal.inject);
                }
                debug.end();
            };

            RoomApplication.prototype.initComponent = function(component) {
                debug.start('initComponent:' + component.name);
                if (!component.registered) {
                    component.registered = true;
                    element.register(this, component.name, component.factory.shadow);
                    (component.factory.init || function() {
                        // Default behavior on registration process
                        debug.info('Registering ' + this.name);
                    }).apply(component, component.inject);
                } else {
                    debug.info('Component ' + component.name + ' already registered');
                }
                debug.end
            }

            RoomApplication.prototype.init = function(hook) {
                var self = this;
                for (var index in self.components) {
                    var component = self.components[index];
                    if (!component.ready && component.required.length > 0) {
                        component.required.forEach(function(subindex) {
                            self.initComponent(self.components[subindex]);
                        });
                    }
                    self.initComponent(component);
                }
                compiler.application(self.name);
                (hook || function() {
                    debug.info('Application ready');
                }).apply(self);
            };

            RoomApplication.prototype.find = function(name) {
                return this.components[util.camelize(name)];
            };

            exports.Application = RoomApplication;

            exports.app = function(name, deps, opts) {
                var inst = new RoomApplication(name, deps, opts);
                return inst;
            };

        }, {
            "./compiler": 12,
            "./dependency": 14,
            "./element": 15,
            "./factory": 16,
            "./util": 22
        }],
        12: [function(require, module, exports) {
            var util = require('./util');
            var config = require('./config');
            var engine = require('mustache');
            var debug = util.debug('Room.compiler', 0);

            exports.application = function(name) {
                var attr = config.prefix + '-' + config.attr.app;
                var appnode = util.$('[' + attr + '="' + util.hyphenate(name) + '"]');
                if (appnode === undefined || !appnode) {
                    var body = document.body;
                    util.$.attr(body, '-' + config.attr.app, util.hyphenate(name));
                }
            };

            exports.render = function(target, data) {
                if (!data || !target) {
                    debug.error('No data or node given for rendering');
                    return false;
                }
                target.innerHTML = engine.render(target.innerHTML, data);
            };

        }, {
            "./config": 13,
            "./util": 22,
            "mustache": 10
        }],
        13: [function(require, module, exports) {
            exports = module.exports = {
                debug: true,
                shadow: true,
                prefix: 'rm',
                attr: {
                    component: 'component',
                    app: 'app'
                }
            };

        }, {}],
        14: [function(require, module, exports) {
            var debug = require('./util').debug('Dependency', 0);

            exports.inject = function(app, name) {
                debug.start('inject:' + name);
                var result = app.components[name];
                if (result) {
                    return result;
                }
                app.dependencies.forEach(function(dependency) {
                    debug.info('Looking for global dependency ' + dependency);
                    if (dependency === name) {
                        debug.info('Global dependency found!');
                        result = window[dependency];
                    }
                });
                return result;
                debug.end();
            };

        }, {
            "./util": 22
        }],
        15: [function(require, module, exports) {
            var util = require('./util');
            var debug = util.debug('Element', -1);
            var config = require('./config');

            var registeredByDefault = [
                'header',
                'div',
                'footer',
                'main',
                'p',
                'span',
                'nav',
                'ul',
                'dl',
                'a',
                'i'
            ];

            var isRegisteredElement = function(name) {
                return document.createElement(name).constructor !== HTMLElement;
            };

            exports.isRegistered = isRegisteredElement;

            exports.register = function(app, identifier, hook) {
                debug.start('register');
                if (isRegisteredElement(identifier) || registeredByDefault.indexOf(identifier) > -1) {
                    debug.info('Element ' + identifier + ' already registered in the DOM');
                    return false;
                }
                var proto = Object.create(HTMLElement.prototype);
                proto.createdCallback = function() {
                    var shadow = this.createShadowRoot();
                    (hook ||   function() {}).apply(shadow);
                };
                var anchestor = app.elements[util.camelize(identifier)];
                if (anchestor === undefined || !anchestor) {
                    anchestor = document.registerElement(util.hyphenate(identifier), {
                        prototype: proto
                    });
                }
                debug.end();
            };

        }, {
            "./config": 13,
            "./util": 22
        }],
        16: [function(require, module, exports) {
            var config = require('./config');
            var $$ = require('./util/query').$;
            var compiler = require('./compiler');
            var hyphenate = require('./util/core').hyphenate;
            var camelize = require('./util/core').camelize;
            var listen = {
                on: require('./util/dom').on,
                off: require('./util/dom').off
            };

            exports.component = function(factory) {
                factory.fire = function fire(event, data, hook) {
                    var target = 'on' + event.charAt(0).toUpperCase() + event.slice(1);
                    (hook || function() {})(self.factory[target].apply(self, data));
                };

                factory.get = function getNodes() {
                    var all = [],
                        final = [],
                        target = hyphenate(factory.name);
                    [
                        $$('.' + target),
                        $$('[' + config.prefix + '-' + config.attr.component + '="' + target + '"]'),
                        $$('@' + target)
                    ].forEach(function(node) {
                        if (typeof node !== undefined && node) {
                            all.push(node);
                        }
                    });
                    all.forEach(function(set) {
                        if (set && typeof set !== undefined) {
                            try {
                                [].slice.call(set).forEach(function(child) {
                                    final.push(child);
                                });
                            } catch (ex) {
                                console.error(ex.message);
                            }
                        }
                    });
                    return final;
                };

                factory.bind = function bindEvent(event, hook) {
                    factory.nodes.forEach(function(finding) {
                        if (typeof finding !== undefined && finding !== null) {
                            listen.on(finding, event, hook);
                        }
                    });
                };

                factory.renderData = function renderData(data) {
                    this.viewbag = data;
                    factory.nodes.forEach(function(node) {
                        compiler.render(node, data);
                    });
                };

                factory.unbind = function unbindEvent(event, hook) {
                    factory.nodes.forEach(function(finding) {
                        if (typeof finding !== undefined && finding !== null) {
                            listen.off(finding, event, hook);
                        }
                    });
                };

                factory.nodes = factory.get();

                return factory;
            };

        }, {
            "./compiler": 12,
            "./config": 13,
            "./util/core": 19,
            "./util/dom": 21,
            "./util/query": 23
        }],
        17: [function(require, module, exports) {
            (function(process) {
                var util = require('./util');
                var debug = util.debug('Room', 0)._env('node');

                if (typeof window === 'undefined' && process) {
                    debug.error('Room must run in a browser and not Node.js');
                }

                var room = room || {};
                room.app = require('./app').app;
                room.$ = require('./util/query').$;
                util.mixin(room.$, require('./util/ajax'));
                exports = module.exports = room;

            }).call(this, require('_process'))

        }, {
            "./app": 11,
            "./util": 22,
            "./util/ajax": 18,
            "./util/query": 23,
            "_process": 1
        }],
        18: [function(require, module, exports) {
            var debug = require('./debug')('Room.ajax');
            exports.get = function(url, callback) {
                var xhr;

                if (typeof XMLHttpRequest !== 'undefined') {
                    xhr = new XMLHttpRequest();
                } else {
                    var versions = [
                        "MSXML2.XmlHttp.5.0",
                        "MSXML2.XmlHttp.4.0",
                        "MSXML2.XmlHttp.3.0",
                        "MSXML2.XmlHttp.2.0",
                        "Microsoft.XmlHttp"
                    ];

                    for (var i = 0, len = versions.length; i < len; i++) {
                        try {
                            xhr = new ActiveXObject(versions[i]);
                            break;
                        } catch (ex) {
                            debug.error(ex.message);
                        }
                    }
                }

                xhr.onreadystatechange = function() {
                    if (xhr.readyState < 4 ||  xhr.status !== 200) {
                        debug.warn('XMLHttpRequest not ready yet');
                        return;
                    }

                    if (xhr.readyState === 4) {
                        callback(xhr);
                    }
                };

                xhr.open('GET', url, true);
                xhr.send('');
            }

        }, {
            "./debug": 20
        }],
        19: [function(require, module, exports) {
            /**
             * Checks if a string starts with '$' or underscore '_'
             * @param {String} val      Input
             * @return {Boolean}
             */
            exports.isReserved = function(val) {
                var c = (val + '').charCodeAt(0);
                return c === 0x24 || c === 0x5F;
            }

            /**
             * Transform a string to uppercase
             * @param  {String} c
             * @return {String}
             */
            function toUpper(_, c) {
                return c ? c.toUpperCase() : '';
            }

            /**
             * Camelize a hyphen-delmited string.
             * @param {String} str      Hyphed Input
             * @return {String}
             */
            exports.camelize = function(hyphed) {
                return hyphed.replace(/-(\w)/g, toUpper);
            }

            /**
             * Hyphenate a camelCase string.
             * @param {String} camel    Camelcase Input
             * @return {String}
             */
            exports.hyphenate = function(camel) {
                return camel
                    .replace(/([a-z\d])([A-Z])/g, '$1-$2')
                    .toLowerCase();
            }

            /**
             * Simple bindings even faster than native implementation
             * @param {Function} fn
             * @param {Object} ctx
             * @return {Function}
             */
            exports.bind = function(handle, context) {
                return function(a) {
                    var size = arguments.length
                    return size ? size > 1 ? handle.apply(context, arguments) : handle.call(context, a) : handle.call(context);
                };
            }

            /**
             * Mix properties into target object.
             * @param {Object} to
             * @param {Object} from
             */
            exports.mixin = function(to, from) {
                for (var key in from) {
                    to[key] = from[key]
                }
                return to;
            }

            /**
             * Check if a needle exists in an array
             * @param  {Array} arr
             * @param  {String} needle
             * @return {Boolean}
             */
            exports.contains = function(arr, needle) {
                for (var i in arr) {
                    if (arr[i] === needle) {
                        return true;
                    }
                }
                return false;
            }

        }, {}],
        20: [function(require, module, exports) {
            var chalk = require('chalk');

            /**
             * Easy debugging wrapped-up in a class,
             * set loglevel with the second parameter;
             *
             *   -1  Silent mode (hide everything)
             *    0  Display only errors
             *    1  Display warnings and errors
             *    2  Display warnings, errors and infos
             *    3  Display everything
             *
             * @param {String} namespace
             * @param {Integer} level
             */
            function RoomDebugger(namespace, level) {
                this.namespace = namespace || '<unknown>';
                this.enabled = true;
                this.level = level || 0;
                this.env = 'browser';
                this.writers = {
                    browser: {
                        log: function(str) {
                            if (level <= 3) {
                                return;
                            }
                            console.log(str);
                        },
                        error: function(str) {
                            if (level <= 0) {
                                return;
                            }
                            console.error(str);
                        },
                        warn: function(str) {
                            if (level <= 1) {
                                return;
                            }
                            console.warn(str);
                        },
                        info: function(str) {
                            if (level <= 2) {
                                return;
                            }
                            console.info(str);
                        },
                        success: function(str) {
                            if (level <= 3) {
                                return;
                            }
                            console.succes(str);
                        }
                    },
                    node: {
                        log: function(str) {
                            console.log(str);
                        },
                        error: function(str) {
                            console.log(chalk.red(str));
                        },
                        warn: function(str) {
                            console.log(chalk.yellow(str));
                        },
                        info: function(str) {
                            console.log(chalk.cyan(str));
                        },
                        success: function(str) {
                            console.log(chalk.green(str));
                        }
                    }
                }
            }

            RoomDebugger.prototype._env = function(environement) {
                this.env = environement;
            };

            RoomDebugger.prototype._out = function(mode, message) {
                if (this.enabled && this.level > -1) {
                    var stdout = this.writers[this.env][mode];
                    if (typeof stdout === 'function') {
                        stdout('[' + this.namespace + '] ' + message);
                    } else {
                        try {
                            stdout.apply(console, '[' + this.namespace + '] ' + message);
                        } catch (ex) {
                            throw new Error(ex);
                        }
                    }
                }
            };

            RoomDebugger.prototype.start = function(subspace) {
                if (console.group) {
                    if (this.level > 0) {
                        console.group(this.namespace + (subspace ? '.' + subspace : ''));
                    }
                }
            };

            RoomDebugger.prototype.end = function() {
                if (console.groupEnd) {
                    if (this.level > 0) {
                        console.groupEnd();
                    }
                }
            };

            RoomDebugger.prototype.log = function(message) {
                this._out('log', message);
            };

            RoomDebugger.prototype.warn = function(message) {
                this._out('warn', message);
            };

            RoomDebugger.prototype.info = function(message) {
                this._out('info', message);
            };

            RoomDebugger.prototype.error = function(message) {
                this._out('error', message);
            };

            RoomDebugger.prototype.success = function(message) {
                this._out('success', message);
            };

            exports = module.exports = function(namespace, level) {
                return new RoomDebugger(namespace, level);
            };

        }, {
            "chalk": 2
        }],
        21: [function(require, module, exports) {
            var config = require('./../config');

            /**
             * Getting and setting attributes on a node
             * @param  {Node} node
             * @param  {String} key
             * @param  {String} val
             * @return {?String}
             */
            exports.attr = function(node, key, val) {
                key = config.prefix + key;
                if (!val && node) {
                    return node.getAttribute(key);
                } else if (node && key && val) {
                    return node.setAttribute(key, val);
                }
            };

            /**
             * Insert node before target
             * @param {Node} node
             * @param {Node} target
             */
            exports.before = function(node, target) {
                target.parentNode.insertBefore(node, target);
            }

            /**
             * Insert node after target
             * @param {Node} node
             * @param {Node} target
             */
            exports.after = function(node, target) {
                if (target.nextSibling) {
                    exports.before(node, target.nextSibling);
                } else {
                    target.parentNode.appendChild(node);
                }
            }

            /**
             * Remove node from DOM
             * @param {Node} node
             */
            exports.remove = function(node) {
                node.parentNode.removeChild(node);
            }

            /**
             * Prepend node to target
             *
             * @param {Node} node
             * @param {Node} target
             */

            exports.prepend = function(node, target) {
                if (target.firstChild) {
                    exports.before(node, target.firstChild);
                } else {
                    target.appendChild(node);
                }
            }

            /**
             * Replace target with node
             * @param {Node} target
             * @param {Node} node
             */
            exports.replace = function(target, node) {
                var parent = target.parentNode;
                if (parent) {
                    parent.replaceChild(node, target);
                }
            }

            exports.on = function(node, event, handler) {
                if (node.addEventListener) {
                    node.addEventListener(event, handler, false);
                } else if (node.attachEvent) {
                    node.attachEvent(event, handler);
                }
            }


            /**
             * Remove event listener shorthand
             * @param {Node} node
             * @param {String} event
             * @param {Function} handler
             */
            exports.off = function(node, event, handler) {
                if (node.removeEventListener) {
                    node.removeEventListener(event, handler);
                } else if (node.detachEvent) {
                    node.detachEvent(event, handler);
                }
            }

            /**
             * Add class with compatibility for IE & SVG
             * @param {Node} node
             * @param {Strong} cls
             */
            exports.addClass = function(node, cls) {
                if (node.classList) {
                    node.classList.add(cls);
                } else {
                    var cur = ' ' + (node.getAttribute('class') || '') + ' ';
                    if (cur.indexOf(' ' + cls + ' ') < 0) {
                        node.setAttribute('class', (cur + cls).trim());
                    }
                }
            }

            /**
             * Remove class with compatibility for IE & SVG
             * @param {Node} node
             * @param {Strong} cls
             */
            exports.removeClass = function(node, cls) {
                if (node.classList) {
                    node.classList.remove(cls);
                } else {
                    var cur = ' ' + (node.getAttribute('class') || '') + ' ';
                    var tar = ' ' + cls + ' ';
                    while (cur.indexOf(tar) >= 0) {
                        cur = cur.replace(tar, ' ');
                    }
                    node.setAttribute('class', cur.trim());
                }
            }

        }, {
            "./../config": 13
        }],
        22: [function(require, module, exports) {
            var core = require('./core');
            core.mixin(exports, core);
            core.mixin(exports, require('./query'));
            core.mixin(exports.$, require('./dom'));
            exports.debug = require('./debug');
            exports.ajax = require('./ajax');

        }, {
            "./ajax": 18,
            "./core": 19,
            "./debug": 20,
            "./dom": 21,
            "./query": 23
        }],
        23: [function(require, module, exports) {
            /**
             * Custom and mini-implementation of jQuery selectors
             * which helps to select nodes in the browser
             * @param  {string} selector    CSS2/3 Selector
             * @return {Node}               0, 1 or multiple Nodes
             */
            var query = function(selector) {
                var matches = {
                        '#': 'getElementById', // $('#myId')
                        '.': 'getElementsByClassName', // $('.myClass')
                        '@': 'getElementsByName', // $('@myName')
                        '=': 'getElementsByTagName', // $('=body')
                        '?': 'querySelectorAll' // $('?anything')
                    },
                    rex = /[=#@.*]/.exec(selector)[0];
                var nodes = document[matches[rex]](selector.split(rex)[1]);
                return nodes.length > 1 ? nodes : nodes[0];
            };

            exports.$ = query;

        }, {}]
    }, {}, [17])(17)
});
